### PR TITLE
Add holder type

### PR DESCRIPTION
**Pull request by muffgaga**

### CODE REVIEW

File: tests/holder_type.cpp [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#include "holder_type.h"
+
+std::shared_ptr<Child> Parent::get_child() {
+  return this->child;
```
Comment (kljohann) [inline tests/holder_type.cpp]:
```suggestion
  return child;
```

File: tests/holder_type.cpp [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+}
+
+long Parent::get_use_count() {
+  return this->child.use_count();
```
Comment (kljohann) [inline tests/holder_type.cpp]:
```suggestion
  return child.use_count();
```

File: genpybind/decls/klasses.py [hunk 3]
```diff
@@ -70,6 +72,15 @@ def set_inline_base(self, *patterns):
...
         self._inline_base.update(patterns)
 
+    @property
+    def holder_type(self):
+        # type: () -> FrozenSet[Text]
+        return frozenset(self._holder_type)
+
+    def set_holder_type(self, *patterns):
+        # type: (*Text) -> None
+        self._holder_type.update(patterns)
```
Comment (kljohann) [inline genpybind/decls/klasses.py]: There can only be a single holder type and using this attribute without arguments should probably be an error. Can you try the following and change the `holder_type` / `statements` functions accordingly?
```python
    def set_holder_type(self, holder_type):
        # type: Optional[Text] -> None
        self._holder_type = holder_type
```

File: genpybind/decls/klasses.py [hunk 4]
```diff
@@ -31,6 +32,7 @@ def __init__(self, *args, **kwargs):
         self._dynamic_attr = False
         self._hide_base = set()  # type: Set[Text]
         self._inline_base = set()  # type: Set[Text]
+        self._holder_type = set()  # type: Set[Text]
```
Comment (kljohann) [inline genpybind/decls/klasses.py]:
```suggestion
        self._holder_type = None  # type: Optional[Text]
```
(see below)

File: genpybind/decls/klasses.py [hunk 5]
```diff
@@ -119,7 +130,8 @@ def base_specifiers(self):
                 bases.append(child.type.fully_qualified_name)
                 continue
             bases.extend(child.base_specifiers())
-
+        filter_bases = ['std::enable_shared_from_this']
```
Comment (kljohann) [inline genpybind/decls/klasses.py]: This somewhat duplicates the `hide_base` functionality. Though I agree that we probably always want to hide `enable_shared_from_this`. Can you integrate this into the `bases` function?

File: genpybind/decls/klasses.py [hunk 6]
```diff
@@ -189,7 +201,7 @@ def statements(self, parent, registry):
 
         yield "auto {var} = py::class_<{classes}>({args});".format(
             var=var,
-            classes=", ".join([self.fully_qualified_name] + self.base_specifiers()),
+            classes=", ".join([self.fully_qualified_name] + self.base_specifiers() + list(self.holder_type)),
```
Comment (kljohann) [inline genpybind/decls/klasses.py]: Maybe add a comment that the order of extra arguments does not matter. As the line is getting rather long, you could also extract this to a variable.

```python
        # Wrapped class has to be specified first; order of extra arguments does not matter
        options = [self.fully_qualified_name] + self.base_specifiers()
        if self.holder_type is not None:
            options.append(self.holder_type)

        # ...

            classes=", ".join(options),
```

File: tests/holder_type.h [hunk 7]
```diff
@@ -0,0 +1,18 @@
+#pragma once
+
+#include <memory>
+
+#include "genpybind.h"
+
+class GENPYBIND(visible, holder_type("std::shared_ptr<Child>")) Child : public std::enable_shared_from_this<Child> {};
```
Comment (kljohann) [inline tests/holder_type.h]: Can you run `clang-format` on both `holder_type.h` and `holder_type.cpp`?

**Review by kljohann — CHANGES_REQUESTED** [review]
Thank you for the pull request and sorry that I noticed it just now (I did not get a notification). Some minor comments inline.

**Comment by muffgaga** [comment]
ACK → @yungwundi, please fix :)

**Comment by timowunderlich** [comment]
@kljohann fixed, hopefully!

File: genpybind/decls/klasses.py [hunk 8]
```diff
@@ -23,14 +23,18 @@ class Klass(Level):
...
     )
 
+    filter_bases = ['*std::enable_shared_from_this*']
+
     def __init__(self, *args, **kwargs):
         # type: (*Any, **Any) -> None
         super(Klass, self).__init__(*args, **kwargs)
         self._dynamic_attr = False
-        self._hide_base = set()  # type: Set[Text]
+        self._hide_base = set(Klass.filter_bases)  # type: Set[Text]
```
Comment (muffgaga) [inline genpybind/decls/klasses.py:35]: ~~@yungwundi can you comment on this change?~~ Did not look at the change in line 29 → ok for me.

File: genpybind/decls/klasses.py [hunk 9]
```diff
@@ -119,7 +132,6 @@ def base_specifiers(self):
                 bases.append(child.type.fully_qualified_name)
                 continue
             bases.extend(child.base_specifiers())
-
```
Comment (muffgaga) [inline genpybind/decls/klasses.py:122]: @yungwundi unrelated change?

File: genpybind/decls/klasses.py (hunk 6 above)
Comment (muffgaga) [reply genpybind/decls/klasses.py]: @yungwundi some additional comment?

**Comment by muffgaga** [comment]
@kljohann Is anything still open (we did some force pushes and the github review view is a bit suboptimal)?

**Review by kljohann — APPROVED** [review]

File: genpybind/decls/klasses.py [hunk 10]
```diff
@@ -70,6 +74,15 @@ def set_inline_base(self, *patterns):
         # type: (*Text) -> None
         self._inline_base.update(patterns)
 
+    @property
+    def holder_type(self):
+        # type: () -> Optional[Text]
+        return self._holder_type
+
+    def set_holder_type(self, holder_type):
+        # type: Optional[Text] -> None
```
Comment (kljohann) [inline genpybind/decls/klasses.py:83]: This has to be `(Optional[Text]) -> None`.

**Comment by kljohann** [comment]
Merged with minor changes (whitespace and Mypy annotation).

**Pull request closed without merging**